libs/state/src/lib/core/operators/distinctUntilSomeChanged.ts

```diff
@@ -0,0 +1,173 @@
+ * @docsCategory operators
+ */
+export function distinctUntilSomeChanged<T extends object>(
+  keyCompareMap: KeyCompareMap<T>
+): MonoTypeOperatorFunction<T>;
+/**
+ * @description
+ *
+ * Returns an Observable that emits all items emitted by the source Observable that are distinct by comparison from
+ * the previous item. Comparison will be done for each set key in the `keys` array.
+ *
+ * If a comparator function is provided, then it will be called for each item to test for whether or not that value should be emitted.
+ * If properties of the source change which are not specified for comparison, no change will be emitted.
+ *
+ * @example
+ * A default comparison for multiple keys.
+ *
+ * ```Typescript
+ * import { of } from 'rxjs';
+ * import { distinctUntilSomeChanged } from 'rx-angular/state';
+ *
+ * interface Person {
+ *    age: number;
+ *    name: string;
+ * }
+ *
+ * of(
+ *   { age: 4, name: 'Foo'},
+ *   { age: 7, name: 'Bar'},
+ *   { age: 5, name: 'Foo'},
+ *   { age: 6, name: 'Foo'},
+ * ).pipe(
+ *   distinctUntilSomeChanged(['age', 'name']),
+ * )
+ * .subscribe(x => console.log(x));
+ *
+ * // displays:
+ * // { age: 4, name: 'Foo' }
+ * // { age: 7, name: 'Bar' }
+ * // { age: 5, name: 'Foo' }
+ * // { age: 6, name: 'Foo' }
+ * ```
+ *
+ * An example with a custom comparison applied to each key
+ * ```Typescript
+ * import { of } from 'rxjs';
+ * import { distinctUntilSomeChanged } from 'rxjs/operators';
+ * import { isDeepEqual } from 'custom/is-equal';
+ *
+ * interface Person {
+ *     age: number;
+ *     name: string;
+ *  }
+ *
+ *  const customCompare = (oldVal, newVal) => isDeepEqual(oldVal, newVal);
+ *
+ *of(
+ *     { age: 4, name: 'Foo1'},
+ *     { age: 7, name: 'Bar'},
+ *     { age: 5, name: 'Foo2'},
+ *     { age: 6, name: 'Foo3'},
+ *   ).pipe(
+ *     distinctUntilSomeChanged(['age', 'name'], customCompare),
+ *   )
+ *   .subscribe(x => console.log(x));
+ *
+ * // displays:
+ * // { age: 4, name: 'Foo1' }
+ * // { age: 7, name: 'Bar' }
+ * // { age: 5, name: 'Foo2' }
+ * // { age: 6, name: 'Foo3' }
+ * ```
+ *
+ * @see {@link distinctUntilChanged}
+ * @see {@link distinctUntilKeyChanged}
+ *
+ * @param {K[]} keys String key for object property lookup on each item.
+ * @param {function} [compare] Optional comparison function called to test if an item is distinct from the previous
+ * item in the source. (applied to each specified key)
+ * @docsPage distinctUntilSomeChanged
+ * @docsCategory operators
+ */
+export function distinctUntilSomeChanged<T extends object, K extends keyof T>(
+  keys: K[],
+  compare?: CompareFn<T[K]>
+): MonoTypeOperatorFunction<T>;
+/**
+ * @internal
+ */
+export function distinctUntilSomeChanged<T extends object, K extends keyof T>(
+  keysOrMap: K[] | KeyCompareMap<T>,
+  compare?: CompareFn<T[K]>
+): MonoTypeOperatorFunction<T> {
+  let distinctCompare: CompareFn<T>;
+  const defaultCompare: CompareFn<T[K]> = (oldVal, newVal) => oldVal === newVal;
```

Comment:
move defaultCompare to top. maybe we find a way to memoize this without the need of any dependency?